File: python/src/kmp.js
class KMP {
  computeLPS = (pattern) => {
    const lps = Array(pattern.length);
    lps[0] = 0;
    for (let i = 1; i < pattern.length; i++) {
      let matched = lps[i - 1];
      while (matched > 0 && pattern[i] !== pattern[matched]) {
        matched = lps[matched - 1];
      }
      if (pattern[i] === pattern[matched]) {
        matched++;
      }
      lps[i] = matched;
    }
    return lps;
  };

  KMPSearch = (text, pattern) => {
    if (!pattern || !text) {
      return []; // no results
    }

    // lps[i] = length of proper prefix of pattern[0]...pattern[i-1]
    //          which is also proper suffix of it
    const lps = this.computeLPS(pattern);
    const result = [];

    let matched = 0;
    for (let i = 0; i < text.length; i++) {
      while (matched > 0 && text[i] !== pattern[matched]) {
        matched = lps[matched - 1];
      }
      if (text[i] === pattern[matched]) {
        matched++;
      }
      if (matched === pattern.length) {
        result.push(i - pattern.length + 1);
        matched = lps[matched - 1];
      }
    }

    return result;
  };
}

export default KMP;

File: python/src/kmp.py
class KMP:
  def computeLPS(self, pattern):
    lps = [0] * len(pattern)
    length = 0
    i = 1
    while i < len(pattern):
      if pattern[i] == pattern[length]:
        length += 1
        lps[i] = length
        i += 1
      else:
        if length != 0:
          length = lps[length - 1]
        else:
          lps[i] = 0
          i += 1
    return lps

  def KMPSearch(self, text, pattern):
    if not pattern or not text:
      return []  # no results

    lps = self.computeLPS(pattern)
    result = []

    i = 0  # index for text
    j = 0  # index for pattern
    while i < len(text):
      if pattern[j] == text[i]:
        i += 1
        j += 1

      if j == len(pattern):
        result.append(i - j)
        j = lps[j - 1]
      elif i < len(text) and pattern[j] != text[i]:
        if j != 0:
          j = lps[j - 1]
        else:
          i += 1

    return result